design/debug-collector.md

```diff
@@ -0,0 +1,127 @@
+
+Since the debug-collector attempts to pull log data from multiple tracers at once, every iteration, it is possible that some of the tracers are locked. To deal with this, the debug-collector will simply skip the locked tracers. Depending on the logging frequency of tracers, this strategy could occasionally result in waiting multiple iterations to pull a given tracer's logs. Another major downside to this approach is the probe effect that halting the CPU will cause.
+
+### Revised proposal: RaceBuffer
+
+A RaceBuffer is a single-producer, single-consumer, lock free ring buffer. As suggested by the name, the RaceBuffer has data races during operation, but it detects when the data races might have lead to inconsistency. Therefore, it also relies on each read/write to each entry to be atomic. In our case, the entries are ekotrace u32 log entries, and we expect to be running the tracers on 32-bit microcontrollers, so each read/write to a u32 should be atomic.
+
+The main goal behind the buffer is to make it so the reader does not need to write at all. Therefore, the writer does not get any information about actions of the reader. In my prototype implementation, the writer simply writes to a ring buffer, overwriting old data after filling the buffer. The reader periodically performs reads to the buffer, reading all of the new data in the buffer every time a read is performed. It stores the data it reads to an output vector. 
+
+Invariants:
+- In the output vector, every entry is at the index that the writer wrote it at. For instance, the 20th entry that the writer wrote would be the 20th element of the output vector. However, since data in the buffer may be overwritten before the reader reads it, any number of the entries in the output vector may be replaced by a nil value which represents that the entries were missed. 
+- For two-word entries (like clock entries or entries with a payload), the algorithm guarantees that each two word entry will either be fully included or fully omitted in the ouput vector; there will never be half of a two-word entry included.
+
+#### Writer Algorithm
+
+The writer keeps track of a write cursor, which represents the index of the next write. This cursor is incremented every time a write happens, and is not reset when the end of the backing storage is reached. Therefore, the actual index of the backing storage is calculated as: `write_cursor % len_of_storage`.
```

Comment:
Since we're targeting embedded systems, I think it would be reasonable to constrain the ring buffer size to a power of 2. That way we can get rid of the expensive modulo operator and instead use bitwise arithmetic.

Reply:
Yeah, that seems reasonable